src/java.base/share/classes/java/lang/reflect/Executable.java

```diff
@@ -438,13 +441,15 @@ private Parameter[] privateGetParameters() {
     boolean hasRealParameterData() {
         // If this somehow gets called before parameters gets
         // initialized, force it into existence.
+        // This is an acquiring read and must be strictly
+        // before assignment of hasRealParameterData
         if (parameters == null) {
             privateGetParameters();
         }
         return hasRealParameterData;
     }
 
-    private transient volatile boolean hasRealParameterData;
+    private transient boolean hasRealParameterData;
     private transient volatile Parameter[] parameters;
```

Comment:
These can probably be annotated with `@Stable` (see also [GH‑8742]):
```suggestion
    private transient @Stable boolean hasRealParameterData;
    private transient @Stable volatile Parameter[] parameters;
```

[GH‑8742]: https://github.com/openjdk/jdk/pull/8742

Reply:
Another approach would be to keep the parameter array and the boolean in a record, and mark the record field as stable, possibly like what's done in #6889. Keeping them in a record like:
```java
record ParameterData(Parameter[] parameters, boolean real) {}

private transient @Stable ParameterData parameterData;

private ParameterData parameterData() {
    ParameterData data = parameterData;
    if (data != null)
        return data;
    // Create parameter data, cache and return
}

boolean hasRealParameterData() { return parameterData().real(); }
private Parameter[] privateGetParameters() { return parameterData().parameters(); }
```

This code should be more maintainable in the long run as well, with more straightforward concurrency and caching model.

Reply:
If my record approach is used, `hasRealParameterData()` and `privateGetParameters()` should be manually inlined instead, for code clarity. Alternatively, you can rename the components to like `sharedArray`/`sharedParameters`, `isReal` to make the call sites like `parameterData().sharedParameters().clone()` or `parameterData().isReal()` more descriptive.

Currently, `hasRealParameterData` is only used in `Executable::getAllGenericParameterTypes`, and `privateGetParameters` only in `Executable::getParameters` (except the unintuitive usage to initialize the parameter data), so the impact of such a refactor and inline would be small.

Reply:
@liach Good point, done!